File: chapter10/color.exs
# color.exs

defmodule Color do
  def from16to24(bitmap) do
    (for << r::5, g::6, b::5 <- bitmap >>, do: <<r::5, 0::3, g::6, 0::2, b::5, 0::3>>) |> Enum.join
  end

  def from24to16(bitmap) do
    (for << r::5, _::3, g::6, _::2, b::5, _::3 <- bitmap >>, do: <<r::5, g::6, b::5>>) |> Enum.join
  end
end

File: chapter10/color.erl
% color.erl

-module(color).
-export([from16to24/1, from24to16/1]).

from16to24(Bitmap) ->
  << <<R:5, 0:3, G:6, 0:2, B:5, 0:3>> || <<R:5, G:6, B:5>> <= Bitmap >>.

from24to16(Bitmap) ->
  << <<R:5, G:6, B:5>> || <<R:5, _:3, G:6, _:2, B:5, _:3>> <= Bitmap >>.
